### DISCUSSION TITLE
0047 XLS-47d Price Oracles

### DISCUSSION THREAD

**Discussion by gregtatcam** [XRPLF/XRPL-Standards#129] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by Silkjaer** [comment]
Serveral existing oracles are using trustlines for the same purpose (e.g. https://github.com/XRPL-Labs/XRPL-Persist-Price-Oracle and https://dev.to/wietse/aggregated-xrp-usd-price-info-on-the-xrp-ledger-1087).

This XLS suggests a new LedgerObject and API methods for the same purpose. The main benefit is less abstraction and better discoverability? Or do I miss something 😀 

IMO on-chain oracles are mainly beneficial to smart contracts (as you also write in the abstract), as off-chain applications might be better served by faster APIs than waiting for off-chain data to be updated to an on-chain oracle to be consumed off-chain again. What is the main purpose for on-chain oracles without smart contract support, or is this envisioned in a Hooks environment?

In that case, it might not even be necessary as Hooks, if I understand correctly, already offers a higher abstraction for storing arbitrary information.

**Reply by jazzi-cooper** [reply]
Thank you, @Silkjaer. I agree with you on the advantages: reduced abstraction, improved discoverability - and generally an enhanced developer experience, particularly beneficial for newcomers to XRPL.

Regarding applications, an oracle becomes essential for potential native DeFi protocols such as lending/borrowing - or possibly the AMM, if adopted. Additionally, there's a need for enhanced price feed infrastructure for explorers and information-focused apps. Hooks, as you noted, will further broaden DeFi opportunities on XRPL, potentially ushering in a range of new use cases.

**Reply by Silkjaer** [reply]
> an oracle becomes essential for potential native DeFi protocols such as lending/borrowing - or possibly the AMM, if adopted. 

Then I think it should be part of one of those native amendments?

> Additionally, there's a need for enhanced price feed infrastructure for explorers and information-focused apps. 

Isn't it faster to use off-chain information, and more reliable? 

> Hooks, as you noted, will further broaden DeFi opportunities on XRPL, potentially ushering in a range of new use cases.

Hooks won't need a native oracle object?

**Reply by jazzi-cooper** [reply]
A dedicated specification for oracles will increase composability, efficiency, and standardization on XRPL. If, say, a lending protocol incorporates its own oracle functionality, it would restrict other protocols, like AMM or other on-chain applications, from tapping into that data without a dependency on lending protocol. 

By setting a standard for price feed data creation, storage, and management, and collaborating with trusted decentralized providers like Chainlink or Band protocol, all XRPL apps and protocols, whether native or hooks-based, can access a dependable shared data source.

**Reply by jazzi-cooper** [reply]
Wallets are another example: The request for in-app pricing feature seems to have been the impetus for this proposal: https://dev.to/wietse/aggregated-xrp-usd-price-info-on-the-xrp-ledger-1087

We would like to formalize this via this proposal for a number of reasons related to improving dev experience, both for data providers and especially data consuming dApps.

**Reply by mvadari** [reply]
> Hooks won't need a native oracle object?

Hooks could still use a native oracle object, right? Since you can just emit a tx. And this way it can be used by other non-Hooks usecases as well, such as integrating with external APIs.
(I might be wrong but I believe that `HookState` objects can only be edited by hooks.)

**Comment by RichardAH** [comment]
Aside from the obvious question "how is this useful without smart contracts?" ... why are _price_ and _scale_ specified individually as uint64 and uint8? The XRPL already has a bespoke base-10 floating point system _STAmount_ which gives you 54 bits of mantissa and 8 bits of exponent. All IOU balances are currently encoded this way, why not use it here too?

**Reply by gregtatcam** [reply]
It doesn't fit well into `STAmount` because there is no issuer account in this case. Also it's common for Price Oracles to provide the price as a scaled uint.

**Comment by dangell7** [comment]
I'm interested in the trust part? Anyone can create prices for any asset? I could see a price, spin up 30 accounts and submit prices?

EDIT: You choose the oracles that you want the price aggregated from. I misunderstood that.

**Reply by dangell7** [reply]
I guess to further, do you see dapps creating their own oracles or do you see dapps using other already created oracles?

**Reply by jazzi-cooper** [reply]
We expect dapps to use third-party oracles from established price feed data furnishers, such as Chainlink, Band, DIA, etc.

Regarding trust, you are correct, we expect dapps to vet the oracles they choose to use by reviewing the creator's site/docs.

**Comment by wojake** [comment]
I like this concept but let's talk use case.

If there were native smart contracts, this would definitely be useful to add to the ledger but what's the rationale behind this without smart contracts as users could *AND* would be requesting price data from third party sites, APIs and services?

**Reply by wojake** [reply]
Without native computation (smart contracts), there is no real (immediate) need for native off-chain data.

**Reply by jazzi-cooper** [reply]
Native computation can be done at the protocol-level without smart contracts - for example the DEX and proposed AMM. Oracles would also be required for a native borrow/lending protocol on XRPL. 

For token-based applications, a native Oracle could be useful for decentralized stablecoins or other collateral-backed assets. Even apps like explorers or token informational sites could benefit from using an on-chain decentralized price data as opposed to centralized APIs, especially if traders are relying on that information to execute trading strategies.

**Comment by shortthefomo** [comment]
Having spent a number of hours actually building Oracles. There are a number of issues I would like to cover with this spec.

1. One ledger entry per Pair is extremely restrictive. If I have 100 pairs I would like data for that now means a 100 records I have to submit.
2. The data is all contained in the ledger there is no optionality. I could very well decide my data should be paid for as a oracle provider.
3. There is no accountability included in the data set that is stored here. How can an oracle provider attest to the data that was submitted? Typically a hash from their ledger or data base should be provided with the submission. So that I as a user can point to the providers data at some point in time and ask for attestation how the result was derived.
4. We already achieve this via an accountSet transaction as done by xumm. It faces the same issues as the data submitted is per pair (https://xrpscan.com/account/rXUMMaPpZqPutoRszR29jtC8amWq3APkx)

Here is an example of data I produce as a Oracle.
<img width="831" alt="Screen Shot 2023-08-18 at 21 35 30" src="https://github.com/XRPLF/XRPL-Standards/assets/508629/153d385f-7405-4a3f-971a-5d2334d35b94">

Notice the vast amount of data I produce with the records, this speaks to the attestation of the result set. The rawData, exchanges data was collected from, timestamp, age of data, etc... From this raw data im able as a provider to attest to the results produced.

I would suggest reworking this spec for the following.
1. Add a URI field to allow off data to be referenced. (optional) 
2. Remove specific fields around a single pair and rather have a single field that allows a limited number of pairs. (if you have more pairs than that its only then an extra transaction to the ledger).
3. Remove this notion of aggregated price from all the oracles unless im able to spec which oracles I want to aggregate price from... this data is much like a trustline you have to trust the provider is not feeding you rubbish.

Here I can now optionally keep my oracle data private (offline to those i service) via the URI. And/Or publicly publish the data via the pair list field.

This makes this a lot more flexible in the way data can be provided from oracles and allows even other ledgers/chains to submit their records here (the URI simply references their hashes) or external centralised databases can be referenced in the same way.

**Reply by gregtatcam** [reply]
Hi @lathanbritz. Thank you for the suggestions.

1. How is off chain data can be used by an on-chain dApp, for instance landing protocol?
2. I have originally considered having an array of pairs in the oracle object but then decided against it because a) the on-ledger object size; b) update transaction complexity as would have to search for each pair in the array. The desired functionality in terms of multiple pair update can be accomplished by changing the proposed update transaction to support array of pairs. The update transaction would look then something like this:
>     {
>        "TransactionType": "UpdateOracle",
>        "Account": "rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW",
>        "prices": [ 
>        {
>            "OracleID": "00070C4495F14B0E44F78A264E41713C64B5F89242540EE255123400000000000000",
>            "SymbolPrice": 540,
>            "Scale": 2,
>            "LastUpdateTime": 743609124
>        },
>       ...
>       {
>          "OracleID": "12370C4495F14B0E44F78A264E41713C64B5F89242540EE255534400000000000000",
>          "SymbolPrice": 740,
>          "Scale": 3,
>          "LastUpdateTime": 743609014
>      }
>     ]
>     }
3. The aggregate price does require specifying the oracle by providing the `OracleID`, which is associated with specific Oracle Provider.

Regarding the attestation. The oracle object includes the Oracle's public key (`Account`), which the Oracle should publish on their website, just like they publish the smart contract address. For instance, this [ETH/USD](https://data.chain.link/ethereum/mainnet/crypto-usd/eth-usd) price feed from chainlink. The transaction must be signed by the `Account` or multi-signers if it is setup for multi-signing. The Oracle provider can publish the hashes of successful update transactions, similarly to the way they publish the smart contract transactions (see the same link above).

**Comment by shortthefomo** [comment]
Hi @gregtatcam 

1. For instance my oracle can return realtime data, the XRPL only closes ledgers every 4 seconds. As open free data that might be perfect for many usecases. If they wanted faster data they could use a connection into the realtime oracle, by me. 
2. This is the trade off I understand, but let's do a small example say we had x5 oracle providers each pushing an average of 50 pairs. That would be 250 transactions every single ledger on top of the normal traffic the XRPL has to handle. This is not sustainable... esp if the number of data providers grows.
3. Missed that super.

Fair enough with regards to using the submission hash as the identifier back to the attestation, agreed that can be achieved like that.

**Reply by gregtatcam** [reply]
Hi @lathanbritz 

1. But this off-chain data can't be used by the ledger, right? First, XRPL doesn't support communicating to an off-chain entity. Even if it did support it then validators would not be able to achieve consensus as they might get a different data because of the latency, etc.
2. On-ledger object must be limited in size. We might have more leeway with the transaction size. As I said, we could batch updates for the same provider into one transaction. So 5 oracles with 50 pairs will result in 5 transactions, not 250.

**Reply by shortthefomo** [reply]
1. Correct, but as XRPL is today there is nothing on chain that would be able to react to this data (that will change i know). But interfaces could opt to use realtime data over ledger data if the provider is trusted. Additionally some L2's eg. Evernode can handle off chain and on chain data, so that consensus is handled there.
2. Right, but the nature of this data is up loaded every ledger so its effect on transaction throughput is much higher than any other type of submission the ledger has had to deal with to date. The balance needs to be both at the size of data on ledger and the traffic to the nodes. Hence again why offering the split between offering RAW price values and/or additional data via a URI is more flexible. Can you point me to how this batching is done I did not see it mentioned here?

**Comment by gregtatcam** [comment]
1. I assume the URI is for each pair, right? So it's probably like 256 bytes. With an array of pairs we are kind of pushing the boundaries. We could increase the reserve requirements depending on the array size. With 50 pairs it's probably going to be one of the largest objects on the ledger.
2. The traffic is going to be about the same, is it not? It might be actually more in case of one oracle with array of pairs since we have to include the asset pair to locate the right entry in the array. And locating the entry is going to have an impact on the transaction throughput since we have to search the array for each pair. The example is in my comment above: https://github.com/XRPLF/XRPL-Standards/discussions/129#discussioncomment-6852281

**Reply by shortthefomo** [reply]
1. One URI per submission is sufficient if it is multiple per submission. Or as you put it one per batch of data.

using your example.. like so
```
{
   "TransactionType": "UpdateOracle",
   "Account": "rsA2LpzuawewSBQXkiju3YQTMzW13pAAdW",
   "URI": "<some-where>",
   "prices": [ 
   {
       "OracleID": "00070C4495F14B0E44F78A264E41713C64B5F89242540EE255123400000000000000",
       "SymbolPrice": 540,
       "Scale": 2,
       "LastUpdateTime": 743609124
   },
  ...
  {
     "OracleID": "12370C4495F14B0E44F78A264E41713C64B5F89242540EE255534400000000000000",
     "SymbolPrice": 740,
     "Scale": 3,
     "LastUpdateTime": 743609014
 }
]
}
```


2. right following now as It did not occur to me with your batch (like example above).

**Reply by gregtatcam** [reply]
1. But URI has to be stored on the ledger, right?

**Reply by shortthefomo** [reply]
1. yes but it not per pair but just for the batch.

**Reply by gregtatcam** [reply]
So it's per oracle and then and it makes a convincing argument (if we include it) to have an array of pairs. Otherwise every oracle would have to have it. I'm still not convinced if we should include it or not based on the current way the ledger works. It could be of course optional, but still if it's used then it might substantially increase the object size. If we do go with the array, we have to reasonably limit the number of pairs. And there is still an issue of a) including in update transaction the asset pair for each price; b) locating the asset pair in the array (though it's not a big deal if it's reasonably small).

**Reply by shortthefomo** [reply]
The URI would need to be at the collection of data submitted. Not the oracle definition. But its not one URI per pair but one URI for  all 50 records (like previous examples)

**Reply by gregtatcam** [reply]
If we move ahead with the array then, I think, we'd have to increase the reserve requirement for Oracle to two. We could have two tiers with one reserve and two reserves requirement. The first tier has the array size of 5, which results in Oracle size ~360. The second tier has the array size of 10, which results in Oracle size ~600.

**Comment by gregtatcam** [comment]
Thank you all for the feedback. I'm going to close this discussion by the end of the next week (October 6) and move the updated specs draft to the standards. The update is going to include the array of token pairs and URI per @lathanbritz suggestion. It'll also consolidate create and update into one transaction.

**Comment by gregtatcam** [comment]
Moved to standards.
